pkg/iac/scanners/helm/parser/parser_tar.go

```diff
@@ -51,61 +52,122 @@ func (p *Parser) addTarToFS(path string) (fs.FS, error) {
 			return nil, fmt.Errorf("failed to get next entry: %w", err)
 		}
 
+		name := filepath.ToSlash(header.Name)
+
 		if checkExistedChart {
 			// Do not add archive files to FS if the chart already exists
 			// This can happen when the source chart is located next to an archived chart (with the `helm package` command)
 			// The first level folder in the archive is equal to the Chart name
-			if _, err := tarFS.Stat(filepath.Dir(path) + "/" + filepath.Dir(header.Name)); err == nil {
+			if _, err := tarFS.Stat(path.Dir(archivePath) + "/" + path.Dir(name)); err == nil {
 				return nil, errSkipFS
 			}
 			checkExistedChart = false
 		}
 
 		// get the individual path and extract to the current directory
-		entryPath := header.Name
+		targetPath := path.Join(path.Dir(archivePath), path.Clean(name))
+
+		link := filepath.ToSlash(header.Linkname)
 
 		switch header.Typeflag {
 		case tar.TypeDir:
-			if err := tarFS.MkdirAll(entryPath, os.FileMode(header.Mode)); err != nil && !errors.Is(err, fs.ErrExist) {
+			if err := tarFS.MkdirAll(targetPath, os.FileMode(header.Mode)); err != nil && !errors.Is(err, fs.ErrExist) {
 				return nil, err
 			}
 		case tar.TypeReg:
-			writePath := filepath.Dir(path) + "/" + entryPath
-			p.debug.Log("Unpacking tar entry %s", writePath)
-
-			_ = tarFS.MkdirAll(filepath.Dir(writePath), fs.ModePerm)
-
-			buf, err := copyChunked(tr, 1024)
-			if err != nil {
+			p.debug.Log("Unpacking tar entry %s", targetPath)
+			if err := copyFile(tarFS, tr, targetPath); err != nil {
 				return nil, err
 			}
-
-			p.debug.Log("writing file contents to %s", writePath)
-			if err := tarFS.WriteFile(writePath, buf.Bytes(), fs.ModePerm); err != nil {
-				return nil, fmt.Errorf("write file error: %w", err)
+		case tar.TypeSymlink:
+			if path.IsAbs(link) {
+				p.debug.Log("Symlink %s is absolute, skipping", link)
+				continue
 			}
+
+			symlinks[targetPath] = path.Join(path.Dir(targetPath), link) // nolint:gosec // virtual file system is used
 		default:
 			return nil, fmt.Errorf("header type %q is not supported", header.Typeflag)
 		}
 	}
 
-	if err := tarFS.Remove(path); err != nil {
-		return nil, fmt.Errorf("failed to remove tar from FS: %w", err)
+	for target, link := range symlinks {
+		if err := copySymlink(tarFS, link, target); err != nil {
+			return nil, fmt.Errorf("copy symlink error: %w", err)
+		}
+	}
+
+	if err := tarFS.Remove(archivePath); err != nil {
+		return nil, fmt.Errorf("remove tar from FS error: %w", err)
 	}
 
 	return tarFS, nil
 }
 
-func copyChunked(src io.Reader, chunkSize int64) (*bytes.Buffer, error) {
-	buf := new(bytes.Buffer)
-	for {
-		if _, err := io.CopyN(buf, src, chunkSize); err != nil {
-			if errors.Is(err, io.EOF) {
-				break
-			}
-			return nil, fmt.Errorf("failed to copy: %w", err)
+func copySymlink(fsys *memoryfs.FS, src, dst string) error {
```

Comment:
Should we take into account recursive symlinks?

Reply:
It's already being handled. Added a test case https://github.com/aquasecurity/trivy/pull/6621/commits/e0bc1ebe759f65dad7b89e11175be0263bca8713